### ISSUE TITLE
[rllib]Does Ray allow for terminal rewards on the last timestep of a simulation?

### ISSUE THREAD

**Issue by Andrej4156** [ray-project/ray#9064]
I have a simulation in which an agent needs to navigate to a point. There are many ways it can fail before arriving (ie falling off a cliff) so in addition to giving rewards each timestep, I've tried giving rewards and penalties at the end of each episode based on how it ended. Unfortunately for some reason these terminal rewards don't have any affect on the agents behavior. Even with very simple rewards (no penalties, bonus reward on arrival at destination) the AI is no more likely to find its way there than it is with just the timestep reward (essentially distance from destination).

Does Ray support these kinds of rewards? Am I doing something wrong or is this just an artifact of this type of learning?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Latest Ray version, Python 3.7, Tensorflow version 2.1, Ubuntu 18.04

**Comment by roireshef** [comment]
RLlib isn't limited to a specific reward function. Completely failing to learn what you seek can be a result of various reasons, as per my experience most of them relate to how you design the MDP - states/actions/transition-distribution/reward-function, and not to the RL algorithm itself. Could you provide your MDP design (environment rules) in more details?

**Comment by Andrej4156** [comment]
The issue is not that it completely fails to learn anything. It's that it can learn the reward function given every step but not rewards given on the last timestep. For proprietary reasons I can't give too many specifics, but I can tell you about the algorithm in general. 

There are many ways the AI can fail. It can run into something, flip over (on both axes), etc. The reward function is just the distance with some exponential functions for its orientation. Basically the more it tilts, the less reward it will get. My problem is that the AI will learn how to get 99% of the way to the goal, and then do something stupid like flip itself over or run into something. It's orientation is a first order system that it directly controls so it's not like something in the environment is forcing it to flip itself. It will repeatedly make the same mistakes at around the same time. Through this all, it learns enough to arrive at the destination up to 80% of the time. Then that percentage drops before it climbs again, never converging. I've posted an example below. Note, the red starts from a pretrained model; that's why it climbs so fast. 

![image](https://user-images.githubusercontent.com/64599107/85244820-ca527000-b413-11ea-8df5-165ca16191d5.png)

The red gives a bonus reward on the last timestep if the AI reaches the destination. The blue does not have a terminal reward, only a timestep reward. And yet it reached the destination just as often as the AI that didn't receive the bonus reward If I had let the blue run it would have dropped similar to the red. The only difference between the two is that the red will have a slightly higher mean reward.

So the AI will learn how to navigate 99% of the episode and I'm trying to get it to consistently learn the last 1% so I can move on to things like minimizing travel time. However, I can't seem to get it to converge to a solution where it arrives at its destination consistently. My thought was to add terminal rewards but that has had no effect on behavior. Even when the AI has no reward but the terminal reward and it starts off reaching the destination, it will eventually stop reaching the destination and "unlearn" everything until it's just making random actions.

**Comment by Andrej4156** [comment]
Update: I've printed the episode rewards through a custom callback function and I can confirm that Ray is receiving these values. I've also been able to determine that if the value is big enough, the reward does affect behavior, just in the opposite way one would expect. So giving the AI a big reward for reaching the destination actually causes it to reach the destination less. Does anybody have recommendations on how to implement terminal rewards to get an agent to consistently reach a destination?
